incomplete_cooperative/regret.py

```diff
@@ -0,0 +1,148 @@
+
+    We minimize regret in every internal node of the game tree.
+    When updating regret, we need to scan through the meta-coalitions from largest to smallest.
+    This identification is refered to as the "rank" of the meta-coalition.
+    Such an order is found using the `metacoalition_ids_by_coalition_size` function, and stored
+    in `meta_id_to_rank`.
+
+    Attributes:
+        number_of_players: number of players in the games.
+        number_of_coalitions: number of coalitions which aren't in K_0.
+        limit_of_revealed: maximum number of coalitions to reveal, i.e. maximum size of metacoal.
+        meta_rank_to_id: map metacoalition size rank -> metacoalition id.
+        meta_id_to_rank: map metacoalition id -> metacoalition size rank.
+        regret: an array of instant regrets of each regret minimizer (zero if an action is inapplicable).
+            This array has size (number_of_regret_minimizers, number_of_coalitions).
+            It is indexed by the meta-coalition ids.
+    """
+
+    def __init__(self, number_of_players: int, limit_of_revealed: int) -> None:
+        """Initialize RM.
+
+        Arguments:
+            number_of_players: number of players in the games.
+            limit_of_revealed: maximum number of coalitions to reveal.
+        """
+        self.number_of_players = number_of_players
+        # number of coalitions we are interested in -- ie not those in K_0
+        self.number_of_coalitions = 2 ** number_of_players - number_of_players - 2
+        self.limit_of_revealed = limit_of_revealed
+
+        # map metacoalition size rank -> metacoalition id
+        self.meta_rank_to_id = metacoalition_ids_by_coalition_size(number_of_players, limit_of_revealed)
+        # these are the metacoalitions of size <= limit_of_revealed
+        self.viable_metacoalitions = len(self.meta_rank_to_id)
+        # map metacoalition id -> metacoalition size rank
+        self.meta_id_to_rank = np.zeros(self.viable_metacoalitions, dtype=int)
+        self.meta_id_to_rank[self.meta_rank_to_id] = np.arange(self.viable_metacoalitions)
+
+        # regret minimizer for every internal node -- except for the bottom layer of game tree
+        self.number_of_regret_minimizers = coalitions_up_to(self.number_of_coalitions, limit_of_revealed - 1)
+
+        self.coalitions_to_player_ids = get_coalition_player_id_map(number_of_players)
+        self.regret = np.zeros((self.number_of_regret_minimizers, self.number_of_coalitions), dtype=RMValue)
+
+    def get_metacoalition_id(self, metacoalition: Iterable[Coalition]) -> int:
+        """Properly get an id of a metacoalition."""
+        coalition_player_ids = self.coalitions_to_player_ids[
+            [x.id for x in metacoalition if len(x) not in [0, 1, 2**self.number_of_players]]]
+        return np.sum(2 ** coalition_player_ids)
+
+    def get_actions_distribution(self, past_actions: Iterable[Coalition] | int) -> np.ndarray:
```

Comment:
Maybe call this `regret_matching` and add `predictive` and `plus` boolean arguments.

Reply:
Changed. The `plus` and `predictive` should I think go to the `__init__`, as we don't really want to use the same regret to minimize oncewith `plus` `True`, and then once with `False`... Makes sense?